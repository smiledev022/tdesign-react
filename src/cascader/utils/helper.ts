import { TreeNode } from '../interface';

/**
 * 获取node节点的全路径
 * @param node
 * @returns
 */
export function getFullPathLabel(node: TreeNode) {
  return node
    .getPath()
    .map((node: TreeNode) => node.label)
    .join('/');
}
