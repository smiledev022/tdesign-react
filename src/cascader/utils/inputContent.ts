import { TreeNode, CascaderContextType, TreeNodeValue } from '../interface';

/**
 * icon Class
 * @param prefix
 * @param CLASSNAMES
 * @param cascaderContext
 * @returns
 */
export function getIconClass(prefix: string, CLASSNAMES: any, cascaderContext: CascaderContextType) {
  const { visible } = cascaderContext;
  return [
    `${prefix}-cascader-icon`,
    {
      [CLASSNAMES.STATUS.visible]: visible,
    },
  ];
}

/**
 * Cascader外显div class
 * @param prefix
 * @param CLASSNAMES
 * @param cascaderContext
 * @returns
 */
export function getCascaderInnerClasses(prefix: string, CLASSNAMES: any, cascaderContext: CascaderContextType) {
  const { disabled, visible, size, multiple } = cascaderContext;
  return [
    `${prefix}-cascader`,
    {
      [CLASSNAMES.STATUS.disabled]: disabled,
      [CLASSNAMES.STATUS.active]: visible,
      [CLASSNAMES.SIZE[size]]: size,
      [`${name}-is-multiple`]: multiple,
    },
  ];
}

/**
 * closeIcon 是否显示
 * @param isHover
 * @param cascaderContext
 * @returns
 */
export function getCloseShow(isHover: boolean, cascaderContext: CascaderContextType) {
  const { multiple, model, disabled, clearable, visible } = cascaderContext;
  return !!(
    !visible &&
    clearable &&
    isHover &&
    !disabled &&
    ((!multiple && model) || (multiple && (model as TreeNodeValue[]).length))
  );
}

/**
 * Placeholder 是否显示
 * @param isHover
 * @param cascaderContext
 * @returns
 */
export function getPlaceholderShow(
  cascaderContext: CascaderContextType,
  singleContent: string,
  multipleContent: TreeNode[],
) {
  const { filterable, visible, multiple } = cascaderContext;
  const valEmpty = !(multiple ? multipleContent.length : singleContent);
  if (filterable) {
    return valEmpty && !visible;
  }
  return valEmpty;
}

/**
 * 单选状态下内容
 * @param isHover
 * @param cascaderContext
 * @returns
 */
export function getSingleContent(cascaderContext: CascaderContextType) {
  const { model, multiple, treeStore, showAllLevels } = cascaderContext;
  if (multiple || !model) return;
  const node = treeStore && treeStore.getNodes(model as TreeNodeValue | TreeNode);
  if (!(node && node.length)) {
    return '';
  }
  const path = node && node[0].getPath();
  if (path && path.length) {
    return showAllLevels ? path.map((node: TreeNode) => node.label).join('/') : path[path.length - 1].label;
  }
  return model as string;
}

/**
 * 多选状态下选中内容
 * @param isHover
 * @param cascaderContext
 * @returns
 */
export function getMultipleContent(cascaderContext: CascaderContextType) {
  const { model, multiple, treeStore } = cascaderContext;

  if (!multiple || !model) return [];
  const node = treeStore && treeStore.getNodes(model as TreeNodeValue | TreeNode);
  if (!node) return [];
  const path = (model as TreeNodeValue[]).map((item: TreeNodeValue) => {
    const node = treeStore.getNodes(item);
    return node[0];
  });
  if (model && (model as TreeNodeValue[]).length) {
    return path && path.length ? path : (model as TreeNode[]);
  }
  return [];
}

/**
 * 点击popup outside副作用
 * @param ref
 * @param cascaderContext
 * @param event
 * @returns
 */
export function outerClickListenerEffect(
  ref: HTMLElement,
  cascaderContext: CascaderContextType,
  event: MouseEvent | TouchEvent,
) {
  const { visible, setVisible } = cascaderContext;
  if (!ref || ref.contains(event.target as Node)) {
    return;
  }

  if (visible) {
    setVisible(false);
  }
}

/**
 * closeIcon点击副作用
 * @param cascaderContext
 */
export function closeIconClickEffect(cascaderContext: CascaderContextType) {
  const { setVisible, setModel, treeStore, multiple } = cascaderContext;

  setModel(multiple ? [] : '');
  treeStore.resetChecked();
  treeStore.resetExpanded();
  setVisible(false);
}

/**
 * tag 关闭按钮点击副作用
 * @param cascaderContext
 */
export function handleRemoveTagEffect(cascaderContext: CascaderContextType, node: TreeNode) {
  const { disabled, setModel } = cascaderContext;

  if (disabled) return;
  const checked = node.setChecked(!node.isChecked());
  setModel(checked);
}

/**
 * 外显content点击副作用
 * @param cascaderContext
 */
export function innerContentClickEffect(cascaderContext: CascaderContextType) {
  const { setVisible, visible, filterActive, disabled } = cascaderContext;

  if (!disabled) {
    if (visible && filterActive) return; // input filter is active
    setVisible(!visible);
  }
}
