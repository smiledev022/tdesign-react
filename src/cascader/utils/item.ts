import { CascaderContextType, TreeNode } from '../interface';

/**
 * 是否显示省略计算方法
 * @param node
 * @param size
 * @returns
 */
export function getLabelIsEllipsis(node: TreeNode, size: CascaderContextType['size']) {
  const sizeMap = {
    small: 11,
    medium: 9,
    large: 8,
  };
  return sizeMap[size] < node.label.length;
}

export function getCascaderItemClass(
  prefix: string,
  node: TreeNode,
  CLASSNAMES: any,
  cascaderContext: CascaderContextType,
) {
  const { checkStrictly, multiple, size } = cascaderContext;
  const expandedActive = !checkStrictly && node.expanded && (multiple ? !node.isLeaf() : true);

  return [
    `${prefix}-cascader-item`,
    {
      [CLASSNAMES.STATUS.selected]: node.checked,
      [CLASSNAMES.STATUS.expanded]: expandedActive,
      [CLASSNAMES.STATUS.disabled]: node.disabled,
      [CLASSNAMES.STATUS.active]: node.actived || (multiple && node.expanded),
      [CLASSNAMES.SIZE[size]]: size,
      [`${prefix}-cascader-item-have-icon`]: node.children,
    },
  ];
}

export function getCascaderItemIconClass(
  prefix: string,
  node: TreeNode,
  CLASSNAMES: any,
  cascaderContext: CascaderContextType,
) {
  const { checkStrictly } = cascaderContext;

  return [
    `${prefix}-cascader-icon`,
    {
      [CLASSNAMES.STATUS.expanded]: !checkStrictly && node.expanded,
    },
  ];
}
