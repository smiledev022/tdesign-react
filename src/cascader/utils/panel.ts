import isFunction from 'lodash/isFunction';
import isNumber from 'lodash/isNumber';
import { TreeNode, CascaderContextType, CascaderProps, ContextType } from '../interface';

/**
 * 面板数据计算方法
 * @param treeNodes
 * @returns
 */
export function getPanels(treeNodes: CascaderContextType['treeNodes']) {
  const panels: TreeNode[][] = [];
  treeNodes.forEach((node: TreeNode) => {
    if (panels[node.level]) {
      panels[node.level].push(node);
    } else {
      panels[node.level] = [node];
    }
  });
  return panels;
}

/**
 * 点击item的副作用
 * @param propsTrigger
 * @param trigger
 * @param node
 * @param cascaderContext
 * @param onChange
 * @param ctx
 */
export function expendClickEffect(
  propsTrigger: CascaderProps['trigger'],
  trigger: CascaderProps['trigger'],
  node: TreeNode,
  cascaderContext: CascaderContextType,
  onChange: CascaderProps['onChange'],
  ctx: any,
) {
  const {
    checkStrictly,
    filterActive,
    multiple,
    treeStore,
    setFilterActive,
    setVisible,
    setModel,
    setTreeNodes,
    setExpend,
  } = cascaderContext;

  // 点击展开节点，设置展开状态
  if (propsTrigger === trigger && !node.isLeaf()) {
    const expanded = node.setExpanded(true);
    treeStore.refreshNodes();
    treeStore.replaceExpanded(expanded);
    const nodes = treeStore.getNodes().filter((node: TreeNode) => node.visible);
    setTreeNodes(nodes);

    // 多选条件下手动维护expend
    if (multiple) {
      setExpend(expanded);
    }
  }

  if (!multiple && (node.isLeaf() || checkStrictly) && trigger === 'click') {
    treeStore.resetChecked();
    const checked = node.setChecked(!node.isChecked());
    const [value] = checked;

    if (onChange && isFunction(onChange)) {
      onChange(value, ctx);
    }

    // 过滤状态下，点击后清除过滤状态
    if (filterActive) {
      setFilterActive(false);
    }

    // 非过滤状态下，关闭
    if (!filterActive) {
      setVisible(false);
    }

    // 非受控状态下更新状态
    setModel(value);
  }
}

/**
 * 多选状态下选中状态数据变化的副作用
 * @param node
 * @param cascaderContext
 * @param onChange
 * @param ctx
 * @returns
 */
export function valueChangeEffect(
  node: TreeNode,
  cascaderContext: CascaderContextType,
  onChange: CascaderProps['onChange'],
  ctx: ContextType,
) {
  const { disabled, max, multiple, setVisible, setModel, filterActive, setFilterActive, treeNodes, treeStore } =
    cascaderContext;

  if (!node || disabled || node.disabled) {
    return;
  }
  const checked = node.setChecked(!node.isChecked());

  if (isNumber(max) && max < 0) {
    console.warn('max should > 0');
  }

  if (checked.length > max && isNumber(max) && max > 0) {
    return;
  }

  if (checked.length === 0) {
    const expanded = treeStore.getExpanded();
    setTimeout(() => {
      treeStore.replaceExpanded(expanded);
      treeStore.refreshNodes();
    }, 0);
  }

  if (!multiple) {
    setVisible(false);
  }

  const isSelectAll = treeNodes.every((item) => checked.indexOf(item.value) > -1);

  if (filterActive && isSelectAll) {
    setVisible(false);
    setFilterActive(false);
  }

  setModel(checked);

  if (onChange && isFunction(onChange)) {
    onChange(checked, ctx);
  }
}
