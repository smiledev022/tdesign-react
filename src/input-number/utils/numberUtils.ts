export const isInValidNumber = (number: number | string) => {
  if (typeof number === 'number') {
    return Number.isNaN(number);
  }

  if (!number) {
    return true;
  }

  return [
    // 普通数字 1.1 1
    /^\s*-?\d+(\.\d+)?\s*$/,
    // 小数点在后 1.
    /^\s*-?\d+\.\s*$/,
    // 小数点在后 .1
    /^\s*-?\.\d+\s*$/,
  ].every((reg) => reg.test(number));
};

export const getNumberPrecision = (number: number | string) => {
  const numStr = String(number);

  return numStr.includes('.') ? numStr.length - numStr.indexOf('.') - 1 : 0;
};

const multiE = (s: string) => {
  const m = s.match(/[e]/gi);
  return m === null ? false : m.length > 1;
};
const multiDot = (s: string) => {
  const m = s.match(/[.]/g);
  return m === null ? false : m.length > 1;
};
const multiNegative = (s: string) => {
  const m = s.match(/[-]/g);
  return m === null ? false : m.length > 1;
};
export const filterInput = (s: string): string => {
  let filterVal = s.replace(/[^\d.eE。-]/g, '').replace('。', '.');
  if (multiE(filterVal) || multiDot(filterVal) || multiNegative(filterVal)) {
    filterVal = filterVal.substr(0, filterVal.length - 1);
  }
  return filterVal;
};
