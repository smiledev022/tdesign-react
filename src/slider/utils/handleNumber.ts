export const numberToPencent = (number: number) => `${number * 100}%`;
