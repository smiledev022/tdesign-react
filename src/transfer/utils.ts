import isEmpty from 'lodash/isEmpty';
import isFunction from 'lodash/isFunction';
import isString from 'lodash/isString';
import React from 'react';
import { TNode } from '../common';

export const getDefaultValue = <T>(value: T | T[], defaultValue?: T): T[] => {
  if (isEmpty(value)) {
    return Array.isArray(defaultValue) ? defaultValue : [defaultValue, defaultValue];
  }

  if (Array.isArray(value)) return value;

  return [defaultValue, defaultValue];
};

export const getJSX = (value: string | TNode) => {
  if (isString(value)) return value;

  if (isFunction(value)) return value();

  if (React.isValidElement(value)) return value;

  return value;
};
