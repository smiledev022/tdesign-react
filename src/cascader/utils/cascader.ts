import { getFullPathLabel } from './helper';
import { TreeNode, TreeNodeValue, TreeOptionData, CascaderContextType } from '../interface';

/**
 * treeValue计算方法
 * @param model
 * @returns
 */
export const getTreeValue = (model: CascaderContextType['model']) => {
  let treeValue: TreeNodeValue[] = [];
  if (Array.isArray(model)) {
    if (model.length > 0 && typeof model[0] === 'object') {
      treeValue = (model as TreeOptionData[]).map((val) => val.value);
    } else if (model.length) {
      treeValue = model as TreeNodeValue[];
    }
  } else if (model) {
    if (typeof model === 'object') {
      treeValue = [(model as TreeOptionData).value];
    } else {
      treeValue = [model];
    }
  }
  return treeValue;
};

/**
 * input和treeStore变化的副作用
 * @param inputVal
 * @param treeStore
 * @param setTreeNodes
 * @returns
 */
export const treeNodesEffect = (
  inputVal: CascaderContextType['inputVal'],
  treeStore: CascaderContextType['treeStore'],
  setTreeNodes: CascaderContextType['setTreeNodes'],
) => {
  if (!treeStore) return;
  let nodes = [];
  if (inputVal) {
    nodes = treeStore.nodes.filter((node: TreeNode) => {
      const fullPathLabel = getFullPathLabel(node);
      return fullPathLabel.indexOf(inputVal) > -1 && node.isLeaf();
    });
  } else {
    nodes = treeStore.getNodes().filter((node: TreeNode) => node.visible);
  }
  setTreeNodes(nodes);
};

/**
 * 初始化展开阶段与展开状态副作用
 * @param treeStore
 * @param treeValue
 * @param expend
 */
export const treeStoreExpendEffect = (
  treeStore: CascaderContextType['treeStore'],
  treeValue: TreeNodeValue[],
  expend: TreeNodeValue[],
) => {
  // init expanded, 无expend状态时设置
  if (Array.isArray(treeValue) && expend.length === 0) {
    const expandedMap = new Map();
    const [val] = treeValue;
    if (val) {
      expandedMap.set(val, true);
      const node = treeStore.getNode(val);
      if (!node) return;
      node.getParents().forEach((tn: TreeNode) => {
        expandedMap.set(tn.value, true);
      });
      const expandedArr = Array.from(expandedMap.keys());
      treeStore.replaceExpanded(expandedArr);
    }
  }
  // 本地维护 expend，更加可控，不需要依赖于 tree 的状态
  if (treeStore.getExpanded() && expend.length) {
    treeStore.replaceExpanded(expend);
  }
  treeStore.refreshNodes();
};
