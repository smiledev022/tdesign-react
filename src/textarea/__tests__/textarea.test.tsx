import React from 'react';
import { testExamples, render, act, waitFor, fireEvent } from '@test/utils';
import userEvent from '@testing-library/user-event';
import { Textarea } from '..';

// 测试组件代码 Example 快照
testExamples(__dirname);

describe('Textarea 组件测试', () => {
  // 测试渲染
  test('create', async () => {
    const { container } = render(<Textarea />);
    expect(container.querySelectorAll('.t-textarea')).not.toBeNull();
  });

  // 测试属性
  test('disabled', async () => {
    const { container } = render(<Textarea disabled={true} />);
    expect(container.querySelectorAll('.t-is-disabled')).not.toBeNull();
  });

  // 测试输入
  test('input', async () => {
    await act(async () => {
      render(<Textarea maxcharacter={5} />);
      // 获取 input
      const inputElement = await waitFor(() => document.querySelector('textarea'));
      expect(inputElement).not.toBeNull();

      const value = 'hello';
      fireEvent.change(inputElement, { target: { value } });
      expect(inputElement.textContent).toBe(value);

      fireEvent.change(inputElement, { target: { value: 'hi,tzmax' } });
      expect(inputElement.textContent.length).toBe(5);
    });

    const onChange = jest.fn();
    const { container } = render(<Textarea maxLength={1} onChange={onChange} />);
    fireEvent.compositionStart(container.querySelector('textarea'));
    fireEvent.change(container.querySelector('textarea'), { target: { value: 'tian' } });
    fireEvent.compositionEnd(container.querySelector('textarea'), {
      currentTarget: { value: '天' },
    });
    fireEvent.change(container.querySelector('textarea'), { target: { value: '天' } });
    expect(onChange).toHaveBeenLastCalledWith('天', expect.objectContaining({}));
  });

  // 测试事件
  test('event', async () => {
    await act(async () => {
      let changeValue = '';
      let event = null;
      render(
        <Textarea
          onChange={(value, e) => {
            changeValue = value;
            event = e;
          }}
          onKeydown={(value, e) => {
            changeValue = value;
            event = e;
          }}
          onKeypress={(value, e) => {
            changeValue = value;
            event = e;
          }}
          onKeyup={(value, e) => {
            changeValue = value;
            event = e;
          }}
        />,
      );
      // 获取 input
      const inputElement = await waitFor(() => document.querySelector('textarea'));
      expect(inputElement).not.toBeNull();

      event = null;
      changeValue = '';
      const value = 'hello';
      fireEvent.change(inputElement, { target: { value } });
      expect(changeValue).not.toBeNull();
      expect(event).not.toBeNull();

      event = null;
      changeValue = '';
      fireEvent.keyDown(inputElement);
      expect(changeValue).not.toBeNull();
      expect(event).not.toBeNull();

      event = null;
      changeValue = '';
      userEvent.type(inputElement, '你好,{enter}World!');
      expect(changeValue).not.toBeNull();
      expect(event).not.toBeNull();

      // fireEvent.change(inputElement, { target: { value: 'hi,tzmax' } });
      // expect(inputElement.textContent.length).toBe(5);
    });
  });
});
