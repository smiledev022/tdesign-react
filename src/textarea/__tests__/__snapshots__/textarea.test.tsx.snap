// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`base.jsx 1`] = `
<DocumentFragment>
  <div
    class="t-textarea"
  >
    <textarea
      class="t-textarea__inner"
      placeholder="请输入内容"
    />
  </div>
</DocumentFragment>
`;

exports[`events.jsx 1`] = `
<DocumentFragment>
  <div
    class="t-textarea"
  >
    <textarea
      class="t-textarea__inner"
      placeholder="请输入内容"
    />
  </div>
</DocumentFragment>
`;

exports[`maxlength.jsx 1`] = `
<DocumentFragment>
  <div
    class="t-textarea"
  >
    <textarea
      class="t-textarea__inner t-resize-none"
      maxlength="20"
      placeholder="请输入内容"
    />
    <span
      class="t-textarea__limit"
    >
      0/20
    </span>
  </div>
</DocumentFragment>
`;

exports[`type.jsx 1`] = `
<DocumentFragment>
  <div
    class="tdesign-demo-block-column"
  >
    <div
      class="t-textarea"
    >
      <textarea
        class="t-textarea__inner"
        placeholder="请输入内容"
        readonly=""
      >
        只读状态
      </textarea>
    </div>
    <div
      class="t-textarea"
    >
      <textarea
        class="t-textarea__inner t-is-disabled"
        disabled=""
        placeholder="请输入内容"
      >
        禁用状态
      </textarea>
    </div>
  </div>
</DocumentFragment>
`;
